### PR TITLE
Implement more Get* methods

### PR DESCRIPTION
**Pull request by dataclouder**
* New Get methods for Catalog, AdminCatalog, Vdc, AdminVdc, CatalogItem,
  ExternalNetwork
* Deprecate corresponding functions/methods
* Replace usage of deprecated functions with new methods
* Add entity.go and entity_test.go for semi-generic operations with Get*
  functions
* Move methods for AdminOrg, AdminCatalog, AdminVdc to new files adminorg.go,
  admincatalog.go, adminvdc.go

### CODE REVIEW

File: govcd/entity.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+type genericGetter func(string, bool) (interface{}, error)
+
+// GetEntityByNameOrId finds a generic entity by Name Or ID
+// On success, returns a pointer to the AdminVdc structure and a nil error
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/entity.go]: fixed

File: govcd/user.go [hunk 2]
```diff
@@ -161,21 +161,10 @@ func (adminOrg *AdminOrg) GetUserById(id string, refresh bool) (*OrgUser, error)
...
-	}
-	// If not found by ID, look by name
-	if IsNotFound(byIdErr) {
-		orgUser, byNameErr = adminOrg.GetUserById(identifier, false)
-		return orgUser, byNameErr
-	} else {
-		// On any other error, we return it
-		return nil, byIdErr
-	}
+	byName := func(name string, refresh bool) (interface{}, error) { return adminOrg.GetUserByName(name, refresh) }
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/user.go]: changed

File: govcd/entity_test.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+)
+
+// An interface used to test the result of Get* methods
+type GenericEntity interface {
+	name() string // returns the entity name
+	id() string   // returns the entity ID
+}
+
+// Defines a generic getter to test all Get* methods
+type GetterTestDefinition struct {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/entity_test.go]: changed

File: govcd/entity_test.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+import (
+	"fmt"
+	"reflect"
+
+	. "gopkg.in/check.v1"
+)
+
+// An interface used to test the result of Get* methods
+type GenericEntity interface {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/entity_test.go]: changed

File: govcd/admincatalog.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+// Deletes the Catalog, returning an error if the vCD call fails.
+// Link to API call: https://code.vmware.com/apis/220/vcloud#/doc/doc/operations/DELETE-Catalog.html
+func (adminCatalog *AdminCatalog) Delete(force, recursive bool) error {
+	catalog := NewCatalog(adminCatalog.client)
+	catalog.Catalog = &adminCatalog.AdminCatalog.Catalog
+	return catalog.Delete(force, recursive)
+}
+
+//   Updates the Catalog definition from current Catalog struct contents.
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/admincatalog.go]: Legacy code that maintained the same spacing as the original. Fixed

File: govcd/adminorg.go [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+// GetAdminVdcByName function uses a valid VDC name and returns a admin VDC object.
+// If no VDC is found, then it returns an empty VDC and no error.
+// Otherwise it returns an empty VDC and an error.
+// Deprecated: Use adminOrg.GetAdminVDCByName
+func (adminOrg *AdminOrg) GetAdminVdcByName(vdcname string) (AdminVdc, error) {
+	for _, vdcs := range adminOrg.AdminOrg.Vdcs.Vdcs {
+		if vdcs.Name == vdcname {
+
+			adminVdc := NewAdminVdc(adminOrg.client)
+
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/adminorg.go]: fixed

File: govcd/adminorg.go [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+// Otherwise it returns an error. Function allows user to use an AdminOrg
+// to also fetch a Catalog. If user does not have proper credentials to
+// perform administrator tasks then function returns an error.
+// API Documentation: https://code.vmware.com/apis/220/vcloud#/doc/doc/operations/GET-Catalog-AdminView.html
+// Deprecated: Use adminOrg.GetAdminCatalog instead
+func (adminOrg *AdminOrg) FindAdminCatalog(catalogName string) (AdminCatalog, error) {
+	for _, catalog := range adminOrg.AdminOrg.Catalogs.Catalog {
+		// Get Catalog HREF
+		if catalog.Name == catalogName {
+
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/adminorg.go]: fixed

File: govcd/entity.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// 	byName := func(name string, refresh bool) (interface{}, error) {
+// 		return org.GetCatalogByName(name, refresh)
+// 	}
+// 	byId := func(id string, refresh bool) (interface{}, error) {
+// 	  return org.GetCatalogById(id, refresh)
+// 	}
+// 	entity, err := GetEntityByNameOrId(byName, byId, identifier, refresh)
+// 	return entity.(*Catalog), err
+// }
+func GetEntityByNameOrId(byName, byId genericGetter, identifier string, refresh bool) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/entity.go]: changed

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,14 +1,25 @@
...
+* Deprecated methods `Catalog.FindCatalogItem`, `Org.FindCatalog`, `Org.GetVdcByName`
+* Deprecated function `GetExternalNetwork`
+* Added methods `Org.GetCatalogByName` and related `Org.GetCatalogById`, `GetCatalogItemByNameOrId`
+* Added methods `VCDClient.GetExternalNetworkByName` and related `GetExternalNetworkById` and `GetExternalNetworkByNameOrId`
+* Added methods `AdminOrg.GetCatalogByName` and related `Org.GetCatalogById`, `GetCatalogByNameOrId`
+* Added methods `AdminOrg.GetAdminCatalogByName` and related `Org.GetAdminCatalogById`, `GetAdminCatalogByNameOrId`
+* Added methods `Org.GetVDCByName` and related `GetVDCById`, `GetVDCByNameOrId`
+* Added methods `AdminOrg.GetVDCByName` and related `GetVDCById`, `GetVDCByNameOrId`
+* Added methods `AdminOrg.GetAdminVDCByName` and related `GetAdminVDCById`, `GetAdminVDCByNameOrId`
+* Added methods `Catalog.Refresh` and `AdminCatalog.Refresh`
```
Comment (lvirbalas) [inline CHANGELOG.md]: It seems to me that this change should be mentioned in the changelog too:
> Move methods for AdminOrg, AdminCatalog, AdminVdc to new files adminorg.go,
admincatalog.go, adminvdc.go

File: CODING_GUIDELINES.md [hunk 10]
```diff
@@ -134,7 +134,7 @@ Each entity should have the following methods:
 // ALWAYS
 (parent *Parent) GetEntityByName(name string) (*Entity, error)
 (parent *Parent) GetEntityById(id string) (*Entity, error)
-(parent *Parent) GetEntityByNameOrId(identifier string) (*Entity, error)
+(parent *Parent) getEntityByNameOrId(identifier string) (*Entity, error)
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply CODING_GUIDELINES.md]: fixed

File: govcd/entity.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+//
+// func (org *Org) GetCatalogByNameOrId(identifier string, refresh bool) (*Catalog, error) {
+// 	getByName := func(name string, refresh bool) (interface{}, error) {
+// 		return org.GetCatalogByName(name, refresh)
+// 	}
+// 	getById := func(id string, refresh bool) (interface{}, error) {
+// 	  return org.GetCatalogById(id, refresh)
+// 	}
+// 	entity, err := getEntityByNameOrId(getByName, getById, identifier, refresh)
+// 	return entity.(*Catalog), err
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply govcd/entity.go]: fixed

File: CHANGELOG.md (hunk 9 above)
Comment (dataclouder) [reply CHANGELOG.md]: added

**Pull request merged**